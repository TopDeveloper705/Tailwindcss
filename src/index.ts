export * from './processor';
export * from './utils';
export * from './config';
export * from './plugin';

module.exports = {
    'colors': require('./config/colors')
}

