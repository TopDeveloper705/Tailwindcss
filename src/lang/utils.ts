export function isSpace(str: string): boolean {
  return /\s/.test(str);
}

export function isDigit(str: string): boolean {
  return /^\d+$/.test(str);
}

export function isAlpha(str: string): boolean {
  return /\w/.test(str);
}

export function connect(strings: string[]): string {
  return strings.join(';\n') + ';';
}

export function findGroupEnd(text: string, startIndex = 0): number {
  let level = 1;
  let endBracket = searchFrom(text, '}', startIndex);
  while (endBracket !== -1) {
    const nextBracket = searchFrom(text, '{', startIndex);
    if (endBracket < nextBracket || nextBracket === -1) {
      level--;
      startIndex = endBracket + 1;
      if (level === 0) return endBracket;
    } else {
      level++;
      startIndex = nextBracket + 1;
    }
    endBracket = searchFrom(text, '}', startIndex);
  }
  return -1;
}

export function searchFrom(
  text: string,
  target: string | RegExp,
  startIndex = 0,
  endIndex?: number
): number {
  // search from partial of string
  const subText = text.substring(startIndex, endIndex);
  const relativeIndex = subText.search(target);
  return relativeIndex === -1 ? -1 : startIndex + relativeIndex;
}
