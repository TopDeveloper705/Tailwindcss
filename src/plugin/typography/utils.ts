export function uniq<T>(array: T[]): T[] {
  return Array.from(new Set(array));
}

export function castArray<T>(value: T): unknown[] {
  if (!arguments.length) {
    return [];
  }
  return Array.isArray(value) ? value : [value];
}

export function isUsableColor(color: string, values: string | {[key:string]:string}): boolean {
  return Boolean(values && typeof values === 'object' && color !== 'gray' && values[600]);
}

export const round = (num: number): string =>
  num
    .toFixed(7)
    .replace(/(\.[0-9]+?)0+$/, '$1')
    .replace(/\.0$/, '');

export const rem = (px: number): string => `${round(px / 16)}rem`;

export const em = (px: number, base: number): string => `${round(px / base)}em`;
