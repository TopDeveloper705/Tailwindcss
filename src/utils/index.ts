export * from "./tools";
export * from "./helpers";
