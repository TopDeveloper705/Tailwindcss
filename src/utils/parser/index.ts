export { default as HTMLParser } from './html';
export { default as CSSParser } from './css';
export { default as ClassParser } from './class';
export { default as SelectorParser } from './selector';
