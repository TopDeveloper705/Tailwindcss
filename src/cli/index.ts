import arg from 'arg';
import { resolve } from 'path';
import { Processor } from '../lib';
import { readFileSync, writeFileSync } from 'fs';
import { HTMLParser } from '../utils/parser';
import { StyleSheet } from '../utils/style';
import {
  getVersion,
  FilePattern,
  walk,
  isFile,
  generateTemplate,
  Console,
} from './utils';

const doc = `
Generate css from text files that containing tailwindcss classes.
By default, it will use interpretation mode to generate a single css file.

Usage:
  windicss [filenames]
  windicss [filenames] -c -m
  windicss [filenames] -c -s -m
  windicss [filenames] [-c | -i] [-b | -s] [-m] [-p <prefix:string>] [-o <path:string>] [--args arguments]

Options:
  -h, --help            Print this help message and exit.
  -v, --version         Print windicss current version and exit.

  -i, --interpret       Interpretation mode, generate class name corresponding to tailwindcss. This is the default behavior.
  -c, --compile         Compilation mode, combine the class name in each row into a single class.
  -t, --preflight       Add preflights, default is false.

  -b, --combine         Combine all css into one single file. This is the default behavior.
  -s, --separate        Generate a separate css file for each input file.

  -m, --minify          Generate minimized css file.
  -p, --prefix PREFIX   Set the css class name prefix, only valid in compilation mode. The default prefix is 'windi-'.
  -o, --output PATH     Set output css file path.
  -f, --config PATH     Set config file path.

  --init PATH           Start a new project on the path.
`;

const args = arg({
  // Types
  '--help': Boolean,
  '--version': Boolean,
  '--compile': Boolean,
  '--interpret': Boolean,
  '--preflight': Boolean,
  '--combine': Boolean,
  '--separate': Boolean,
  '--minify': Boolean,
  '--init': String,
  '--prefix': String,
  '--output': String,
  '--config': String,

  // Aliases
  '-h': '--help',
  '-v': '--version',
  '-i': '--interpret',
  '-c': '--compile',
  '-t': '--preflight',
  '-b': '--combine',
  '-s': '--separate',

  '-m': '--minify',
  '-p': '--prefix',
  '-o': '--output',
  '-f': '--config',
});

if (args['--help'] || (args._.length === 0 && Object.keys(args).length === 1)) {
  Console.log(doc);
  process.exit();
}

if (args['--version']) {
  Console.log(getVersion());
  process.exit();
}

if (args['--init']) {
  const template = generateTemplate(args['--init'], args['--output']);
  args._.push(template.html);
  args['--preflight'] = true;
  args['--output'] = template.css;
}

const localFiles = walk('.', true).filter((i) => i.type === 'file').map(i=> ({ type: i.type, path: i.path.replace(/\\/g, '/') }) );

const matchFiles: string[] = [];
for (const pt of args._) {
  if (isFile(pt) && pt.search(/\.windi\./) === -1) {
    matchFiles.push(pt);
  } else if (pt.search(/\*/) !== -1) {
    // match files like **/*.html **/src/*.html *.html ...
    localFiles.forEach((i) => {
      const pattern = new FilePattern(pt);
      if (pattern.match(i.path) && i.path.search(/\.windi\./) === -1)
        matchFiles.push(i.path);
    });
  } else {
    Console.error(`File ${pt} does not exist!`);
  }
}

let ignoredClasses: string[] = [];
const preflights: StyleSheet[] = [];
const styleSheets: StyleSheet[] = [];
const processor = new Processor(args['--config'] ? require(resolve(args['--config'])) : undefined);

if (args['--compile']) {
  // compilation mode
  const prefix = args['--prefix'] ?? 'windi-';
  matchFiles.forEach((file) => {
    let indexStart = 0;
    const outputStyle: StyleSheet[] = [];
    const outputHTML: string[] = [];
    const html = readFileSync(file).toString();
    const parser = new HTMLParser(html);

    // Match tailwind ClassName then replace with new ClassName
    parser.parseClasses().forEach((p) => {
      outputHTML.push(html.substring(indexStart, p.start));
      const utility = processor.compile(p.result, prefix, true); // Set third argument to false to hide comments;
      outputStyle.push(utility.styleSheet);
      ignoredClasses = [...ignoredClasses, ...utility.ignored];
      outputHTML.push([utility.className, ...utility.ignored].join(' '));
      indexStart = p.end;
    });
    outputHTML.push(html.substring(indexStart));
    styleSheets.push(
      outputStyle.reduce(
        (previousValue: StyleSheet, currentValue: StyleSheet) =>
          previousValue.extend(currentValue),
        new StyleSheet()
      )
    );

    const outputFile = file.replace(/(?=\.\w+$)/, '.windi');
    writeFileSync(outputFile, outputHTML.join(''));
    Console.log(`${file} -> ${outputFile}`);

    if (args['--preflight'])
      preflights.push(processor.preflight(parser.html));
  });
} else {
  // interpretation mode
  matchFiles.forEach((file) => {
    const parser = new HTMLParser(readFileSync(file).toString());
    const utility = processor.interpret(
      parser
        .parseClasses()
        .map((i) => i.result)
        .join(' ')
    );
    styleSheets.push(utility.styleSheet);
    ignoredClasses = [...ignoredClasses, ...utility.ignored];

    if (args['--preflight'])
      preflights.push(processor.preflight(parser.html));
  });
}

if (args['--separate']) {
  styleSheets.forEach((style, index) => {
    const filePath = matchFiles[index].replace(/\.\w+$/, '.windi.css');
    if (args['--preflight']) style = preflights[index].extend(style);
    writeFileSync(filePath, style.build(args['--minify']));
    Console.log(`${matchFiles[index]} -> ${filePath}`);
  });
} else {
  let outputStyle = styleSheets
    .reduce(
      (previousValue: StyleSheet, currentValue: StyleSheet) =>
        previousValue.extend(currentValue),
      new StyleSheet()
    )
    .combine()
    .sort();
  if (args['--preflight'])
    outputStyle = preflights
      .reduce(
        (previousValue: StyleSheet, currentValue: StyleSheet) =>
          previousValue.extend(currentValue),
        new StyleSheet()
      )
      .combine()
      .sort()
      .extend(outputStyle);
  const filePath = args['--output'] ?? 'windi.output.css';
  writeFileSync(filePath, outputStyle.build(args['--minify']));
  Console.log('matched files:', matchFiles);
  Console.log('output file:', filePath);
}

Console.log('ignored classes:', ignoredClasses);
