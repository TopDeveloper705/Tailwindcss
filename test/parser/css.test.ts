import { Processor } from '../../src/lib';
import { CSSParser } from '../../src/utils/parser';

const PROCESSOR = new Processor();
const PARSER = new CSSParser();

describe('CSSParser', () => {
  it('Most Simple', () => {
    const css = 'font-size: 12px;';
    expect(PARSER.parse(css).build()).toEqual('font-size: 12px;');
  });

  it('Single Level', () => {
    const css = `
    img,
    video {
      max-width: 100%;
      height: auto;
    }
    .navigation {
      font-size: 12px;
    }
    .logo {
      width: 300px;
    }
    `;
    expect(PARSER.parse(css).build()).toMatchSnapshot('css');
  });

  it('Nesting', () => {
    const css = `
      #header {
        color: black;
        .navigation {
          font-size: 12px;
        }
        .logo {
          width: 300px;
        }
      }`;
    expect(PARSER.parse(css).build()).toMatchSnapshot('css');
  });

  it('Deep Nesting', () => {
    const css = `
      #header {
        color: black;
        .navigation {
          font-size: 12px;
          .logo {
            width: 300px;
            .abc {
              color: white;
            }
            height: 200px;
          }
        }
      }
    `;
    expect(PARSER.parse(css).build()).toMatchSnapshot('css');
  });

  it('Pseudo Classes', () => {
    const css = `
    #bundle {
      .button {
        display: block;
        border: 1px solid black;
        background-color: grey;
        &:hover {
          background-color: white;
        }
        &::first-line {
          color: red;
        }
        &__abc {
          &:focus {
            color: blue;
          }
        }
      }
      .logo {
        width: 300px;
      }
    }
    `;
    expect(PARSER.parse(css).build()).toMatchSnapshot('css');
  });

  it('One line AtRule', () => {
    const css = `
      @import "library";
      @import "typo.css";
    `;
    expect(PARSER.parse(css).build()).toMatchSnapshot('css');
  });

  it('AtRule wrap properties', () => {
    const css = `
    @font-face {
      font-family: Proxima Nova;
      font-weight: 400;
      src: url(/fonts/proxima-nova/400-regular.woff) format("woff");
    }
    `;
    expect(PARSER.parse(css).build()).toMatchSnapshot('css');
  });

  it('Simple atRule', () => {
    const css = `
    @media (min-width: 768px) {
      .test {
        color: red;
      }
    }
    `;
    expect(PARSER.parse(css).build()).toMatchSnapshot('css');
  });

  it('Nested At-Rules and Bubbling', () => {
    const css = `
      .component {
        width: 300px;
        @media (min-width: 768px) {
          width: 600px;
          @media (min-resolution: 192dpi) {
            background-image: url(/img/retina2x.png);
          }
        }
        @media (min-width: 1280px) {
          width: 800px;
        }
      }
    `;
    expect(PARSER.parse(css).build()).toMatchSnapshot('css');
  });

  it('Nested At-Rules', () => {
    const css = `
      @media (min-width: 768px) {
        .test {
          width: 300px !important;
        }
        @media (min-resolution: 192dpi) {
          .component {
            background-image: url(/img/retina2x.png);
          }
        }
      }
    `;
    expect(PARSER.parse(css).build()).toMatchSnapshot('css');
  });

  it('Keyframes', () => {
    const css = `
    @media (min-width: 640px) {
      @media (prefers-color-scheme: dark) {
        @keyframes ping {
          0% {
            transform: scale(1);
            opacity: 1;
          }
          75%, 100% {
            transform: scale(2);
            opacity: 0;
          }
        }
      }
    }`;
    expect(PARSER.parse(css).build()).toMatchSnapshot('css');
  });

  it('Directives and Functions', () => {
    const css = `
    @screen sm {
      .test {
        padding-top: 1px;
      }
    }

    .container {
      width: 100%;
      margin-right: auto;
      margin-left: auto;
      padding-right: 1rem;
      padding-left: 1rem;
      @apply font-bold lg:bg-green-300 md:text-lg !important;
      @apply text-gray-900;
    }

    @media (min-width: 640px) {
      .container {
        max-width: 640px; /* this is comment */
      }
    }

    @media (min-width: 768px) {
      .container {
        max-width: 768px;
      }
    }

    @screen dark {
      .test {
        @apply bg-gray-300;
      }
    }

    @variants dark {
      .test {
        @apply font-medium;
      }
    }

    @variants lg:hover, focus {
      .button {
        @apply text-green-300;
      }
    }

    .link:hover {
      @apply bg-red-500;
    }

    .link:hover {
      opacity: 1;
    }`;

    const parser = new CSSParser(css, PROCESSOR);
    expect(parser.parse().build()).toMatchSnapshot('css');
  });

  it('apply chain', () => {
    const css = `.base {
      @apply w-9 h-9 rounded-md;
    }
    .extend-base {
      @apply base bg-white;
    }`;
    const parser = new CSSParser(css, PROCESSOR);
    expect(parser.parse().build()).toMatchSnapshot('css');
  });

  it('apply hover with multiple selector', () => {
    const css = `.a, .b {
      @apply hover:text-red-200;
    }`;
    const parser = new CSSParser(css, PROCESSOR);
    expect(parser.parse().build()).toMatchSnapshot('css');
  });

  it('do not change apply order', () => {
    const css = `.test-border {
      @apply border-red-500;
      border-right-color: transparent;
      border-bottom-color: transparent;
      border-left-color: transparent;
    }`;
    const parser = new CSSParser(css, PROCESSOR);
    expect(parser.parse().build()).toMatchSnapshot('css');
  });

  it('theme function', () => {
    const css = `.btn-blue {
      background-color: theme('colors.blue.500');
    }
    .content-area {
      height: calc(100vh - theme('spacing[2.5]', '0.625rem')) + theme('spacing.2', '0.5rem');
    }
    .content-area2 {
      height: calc(100vh - theme('spacing.a', '4rem'));
    }`;
    const parser = new CSSParser(css, PROCESSOR);
    expect(parser.parse().build()).toMatchSnapshot('css');
  });

  it('allow last rule without semicolon', () => {
    const css = '.btn-red {background-color: red  }';
    const parser = new CSSParser(css, PROCESSOR);
    expect(parser.parse().build()).toEqual('.btn-red {\n  background-color: red;\n}');
  });

  it('layer directive', () => {
    const css = `
    @layer components {
      .components {
        @apply bg-red-500;
      }
    }

    @layer utilities {
      .utilities {
        max-width: 768px;
      }
    }

    @layer base {
      base {
        margin-left: auto;
      }
    }

    .normal {
      marign-right: auto; /* components by default */
    }
    `;
    const parser = new CSSParser(css, PROCESSOR);
    expect(parser.parse().sort(true).children.map(i => i.build()).join('\n')).toMatchSnapshot('css');
  });

  it('keeps multiple @font-face', () => {
    const css = `
    @font-face {
      font-family: Proxima Nova;
      font-weight: 400;
      src: url(/fonts/proxima-nova/400-regular.woff) format("woff");
    }
    @font-face {
      font-family: Proxima Nova;
      font-weight: 700;
      src: url(/fonts/proxima-nova/700-regular.woff) format("woff");
    }
    `;
    const parser = new CSSParser(css, PROCESSOR);
    expect(parser.parse(css).build()).toMatchSnapshot('css');
  });

  it('animations with @apply', () => {
    const css = `
    .pulse-class {
      @apply relative w-40 h-40 rounded-full bg-teal-500 opacity-60 animate-pulse;
    }

    @screen dark {
      .ping-class {
        @apply animate-ping;
      }
    }
    `;
    const parser = new CSSParser(css, PROCESSOR);
    expect(parser.parse(css).build()).toMatchSnapshot('css');
  });

  it('comma separated selectors', () => {
    const css = `
    .dark {
      .parent {
        & h2, & h3 {
          & > a {
            &::after {
              color: red;
            }
          }
        }
      }
    }

    .Comp1,
    .Comp2 {
      border: 1px solid red;

      &-child {
        border:10px solid green;
      }
    }

    .alert, .warning {
      ul, p {
        margin-left: 0;
      }
    });


    `;
    const parser = new CSSParser(css, PROCESSOR);
    expect(parser.parse(css).build()).toMatchSnapshot('css');
  });

  it('single line comment', () => {
    const css = `
// comment a
.test {
// comment b
  &-child {
    border: 10px solid green;
    // this is another comment
  }
}
`;
    const parser = new CSSParser(css, PROCESSOR);
    expect(parser.parse(css).build()).toMatchSnapshot('css');
  });

  it('multiple line comment', () => {
    const css = `
    .box {
      border: 10px solid green;
    }
    /* .test {
      &-child {
        border: 10px solid green;
      }
    } */
    `;
    const parser = new CSSParser(css, PROCESSOR);
    expect(parser.parse(css).build()).toMatchSnapshot('css');
  });

  it('last rule without semicolon', () => {
    const css = '.font-bold { @apply font-bold; display: block }';
    const parser = new CSSParser(css, PROCESSOR);
    expect(parser.parse(css).build()).toMatchSnapshot('css');
  });

  it('background-image url', () => {
    const css = `.background {
      background-image: url('https://awebsite.com/an-image.png');
      &__overlay {
        // any rule
        background-color: #023761;
      }
    }`;
    const parser = new CSSParser(css, PROCESSOR);
    expect(parser.parse(css).build()).toMatchSnapshot('css');
  });

  it('@import url', () => {
    const css = `@import url('https://fonts.googleapis.com/css2?family=Barlow:wght@400;500&display=swap');
    #app {
      color: red;
    }`;
    const parser = new CSSParser(css, PROCESSOR);
    expect(parser.parse(css).build()).toMatchSnapshot('css');
  });
});
