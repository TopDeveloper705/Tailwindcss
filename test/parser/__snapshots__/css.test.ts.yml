Tools / @import url / css / 0: >-
  @import
  url('https://fonts.googleapis.com/css2?family=Barlow:wght@400;500&display=swap');

  #app {
    color: red;
  }
Tools / AtRule wrap properties / css / 0: |-
  @font-face {
    font-family: Proxima Nova;
    font-weight: 400;
    src: url(/fonts/proxima-nova/400-regular.woff) format("woff");
  }
Tools / Deep Nesting / css / 0: |-
  #header {
    color: black;
  }
  #header .navigation {
    font-size: 12px;
  }
  #header .navigation .logo {
    width: 300px;
    height: 200px;
  }
  #header .navigation .logo .abc {
    color: white;
  }
Tools / Directives and Functions / css / 0: |-
  .container {
    width: 100%;
    margin-right: auto;
    margin-left: auto;
    padding-right: 1rem;
    padding-left: 1rem;
    font-weight: 700 !important;
    --tw-text-opacity: 1;
    color: rgba(17, 24, 39, var(--tw-text-opacity));
  }
  .dark .test {
    font-weight: 500;
  }
  .button:focus {
    --tw-text-opacity: 1;
    color: rgba(110, 231, 183, var(--tw-text-opacity));
  }
  .link:hover {
    --tw-bg-opacity: 1;
    background-color: rgba(239, 68, 68, var(--tw-bg-opacity));
    opacity: 1;
  }
  @media (min-width: 640px) {
    .test {
      padding-top: 1px;
    }
    .container {
      max-width: 640px;
    }
  }
  @media (min-width: 768px) {
    .container {
      font-size: 1.125rem !important;
      line-height: 1.75rem !important;
      max-width: 768px !important;
    }
  }
  @media (min-width: 1024px) {
    .container {
      --tw-bg-opacity: 1 !important;
      background-color: rgba(110, 231, 183, var(--tw-bg-opacity)) !important;
    }
    .button:hover {
      --tw-text-opacity: 1;
      color: rgba(110, 231, 183, var(--tw-text-opacity));
    }
  }
  @media (prefers-color-scheme: dark) {
    .test {
      --tw-bg-opacity: 1;
      background-color: rgba(209, 213, 219, var(--tw-bg-opacity));
    }
  }
Tools / Keyframes / css / 0: |-
  @media (prefers-color-scheme: dark) {
    @media (min-width: 640px) {
      @keyframes ping {
        0% {
          transform: scale(1);
          opacity: 1;
        }
        75%, 100% {
          transform: scale(2);
          opacity: 0;
        }
      }
    }
  }
Tools / Nested At-Rules / css / 0: |-
  @media (min-width: 768px) {
    .test {
      width: 300px !important;
    }
  }
  @media (min-resolution: 192dpi) {
    @media (min-width: 768px) {
      .component {
        background-image: url(/img/retina2x.png);
      }
    }
  }
Tools / Nested At-Rules and Bubbling / css / 0: |-
  .component {
    width: 300px;
  }
  @media (min-width: 768px) {
    .component {
      width: 600px;
    }
  }
  @media (min-width: 1280px) {
    .component {
      width: 800px;
    }
  }
  @media (min-resolution: 192dpi) {
    @media (min-width: 768px) {
      .component {
        background-image: url(/img/retina2x.png);
      }
    }
  }
Tools / Nesting / css / 0: |-
  #header {
    color: black;
  }
  #header .navigation {
    font-size: 12px;
  }
  #header .logo {
    width: 300px;
  }
Tools / One line AtRule / css / 0: |-
  @import "library";
  @import "typo.css";
Tools / Pseudo Classes / css / 0: |-
  #bundle .button {
    display: block;
    border: 1px solid black;
    background-color: grey;
  }
  #bundle .button:hover {
    background-color: white;
  }
  #bundle .button::first-line {
    color: red;
  }
  #bundle .button__abc:focus {
    color: blue;
  }
  #bundle .logo {
    width: 300px;
  }
Tools / Simple atRule / css / 0: |-
  @media (min-width: 768px) {
    .test {
      color: red;
    }
  }
Tools / Single Level / css / 0: |-
  img, video {
    max-width: 100%;
    height: auto;
  }
  .navigation {
    font-size: 12px;
  }
  .logo {
    width: 300px;
  }
Tools / animations with @apply / css / 0: |-
  @keyframes pulse {
    0%, 100% {
      opacity: 1;
    }
    50% {
      opacity: .5;
    }
  }
  @-webkit-keyframes pulse {
    0%, 100% {
      opacity: 1;
    }
    50% {
      opacity: .5;
    }
  }
  @keyframes ping {
    0% {
      -webkit-transform: scale(1);
      transform: scale(1);
      opacity: 1;
    }
    75%, 100% {
      -webkit-transform: scale(2);
      transform: scale(2);
      opacity: 0;
    }
  }
  @-webkit-keyframes ping {
    0% {
      -webkit-transform: scale(1);
      transform: scale(1);
      opacity: 1;
    }
    75%, 100% {
      -webkit-transform: scale(2);
      transform: scale(2);
      opacity: 0;
    }
  }
  .pulse-class {
    position: relative;
    width: 10rem;
    height: 10rem;
    border-radius: 9999px;
    --tw-bg-opacity: 1;
    background-color: rgba(20, 184, 166, var(--tw-bg-opacity));
    opacity: 0.6;
    -webkit-animation: pulse 2s cubic-bezier(0.4, 0, 0.6, 1) infinite;
    animation: pulse 2s cubic-bezier(0.4, 0, 0.6, 1) infinite;
  }
  @media (prefers-color-scheme: dark) {
    .ping-class {
      -webkit-animation: ping 1s cubic-bezier(0, 0, 0.2, 1) infinite;
      animation: ping 1s cubic-bezier(0, 0, 0.2, 1) infinite;
    }
  }
Tools / apply chain / css / 0: |-
  .base {
    width: 2.25rem;
    height: 2.25rem;
    border-radius: 0.375rem;
  }
  .extend-base {
    width: 2.25rem;
    height: 2.25rem;
    border-radius: 0.375rem;
    --tw-bg-opacity: 1;
    background-color: rgba(255, 255, 255, var(--tw-bg-opacity));
  }
Tools / apply hover with multiple selector / css / 0: |-
  .a:hover, .b:hover {
    --tw-text-opacity: 1;
    color: rgba(254, 202, 202, var(--tw-text-opacity));
  }
Tools / background-image url / css / 0: |-
  .background {
    background-image: url('https://awebsite.com/an-image.png');
  }
  .background__overlay {
    background-color: #023761;
  }
Tools / comma separated selectors / css / 0: |-
  .dark .parent h2 > a::after, .dark .parent h3 > a::after {
    color: red;
  }
  .Comp1, .Comp2 {
    border: 1px solid red;
  }
  .Comp1-child, .Comp2-child {
    border: 10px solid green;
  }
  .alert ul, .alert p, .warning ul, .warning p {
    margin-left: 0;
  }
Tools / do not change apply order / css / 0: |-
  .test-border {
    --tw-border-opacity: 1;
    border-color: rgba(239, 68, 68, var(--tw-border-opacity));
    border-right-color: transparent;
    border-bottom-color: transparent;
    border-left-color: transparent;
  }
Tools / keeps multiple @font-face / css / 0: |-
  @font-face {
    font-family: Proxima Nova;
    font-weight: 400;
    src: url(/fonts/proxima-nova/400-regular.woff) format("woff");
  }
  @font-face {
    font-family: Proxima Nova;
    font-weight: 700;
    src: url(/fonts/proxima-nova/700-regular.woff) format("woff");
  }
Tools / last rule without semicolon / css / 0: |-
  .font-bold {
    font-weight: 700;
    display: block;
  }
Tools / layer directive / css / 0: |-
  base {
    margin-left: auto;
  }
  .components {
    --tw-bg-opacity: 1;
    background-color: rgba(239, 68, 68, var(--tw-bg-opacity));
  }
  .normal {
    marign-right: auto;
  }
  .utilities {
    max-width: 768px;
  }
Tools / multiple line comment / css / 0: |-
  .box {
    border: 10px solid green;
  }
Tools / single line comment / css / 0: |-
  .test-child {
    border: 10px solid green;
  }
Tools / theme function / css / 0: |-
  .btn-blue {
    background-color: #3b82f6;
  }
  .content-area {
    height: calc(100vh - 0.625rem) + 0.5rem;
  }
  .content-area2 {
    height: calc(100vh - 4rem);
  }
